### PR TITLE
Implement modified key export API for Mbed TLS 3.0 

### LINKED ISSUE #4363: Unify SSL key exporter API and prepare for TLS 1.3
__Context:__ Some time ago, the original key exporter API `mbedtls_ssl_conf_export_keys_cb()` has been found insufficient and extended as `mbedtls_ssl_conf_export_keys_ext_cb()`, adding the ClientHello random value as a session identifier.

__Task:__ Use Mbed TLS 3.0 to unify the key exporter API, agreeing on and implementing a single exporter callback type and API.

Ideally, this should take suitability for TLS 1.3 into account, which adds a number of new keys to be exported.

### PR DESCRIPTION
**Pull request by hanno-arm**
This resolves #4363 and implements the API proposed therein. See #4363 and the documentation for more information.

### CODE REVIEW

**Comment by hanno-arm** [comment]
CI only failing because of the API break.

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -976,6 +964,58 @@ struct mbedtls_ssl_session
...
+ * \brief           Callback type: Export key alongside random values for
+ *                                 session identification, and PRF for
+ *                                 implementation of TLS key exporters.
+ *
+ * \param p_expkey  Context for the callback.
+ * \param type      The type of the key that is being exported.
+ * \param client_random The client random bytes.
+ * \param server_random The server random bytes.
+ * \param tls_prf_type The identifier for the PRF used in the handshake
+ *                     to which the key belongs.
```
Comment (daverodgman) [inline include/mbedtls/ssl.h:1005]: Missing docs for secret and secret_len

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -976,6 +964,58 @@ struct mbedtls_ssl_session
 #endif
 };
 
+/*
+ * Identifiers for PRFs used in various versions of TLS.
+ */
+typedef enum
+{
+   MBEDTLS_SSL_TLS_PRF_NONE,
+   MBEDTLS_SSL_TLS_PRF_TLS1,
```
Comment (daverodgman) [inline include/mbedtls/ssl.h:973]: Is MBEDTLS_SSL_TLS_PRF_TLS1 still needed after TLS 1.0 and 1.1 have been removed?

**Review by daverodgman — CHANGES_REQUESTED** [review]
A couple of minor points, otherwise looks good

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h:973]: Yes :+1:

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h:1005]: Fixing

**Comment by hanno-arm** [comment]
@daverodgman Thank you for your review. I believe to have addressed your comments. Could you please re-review?

**Comment by hanno-arm** [comment]
@ronald-cron-arm Why was this removed from the Mbed TLS 3.0 Optional EPIC?

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: @daverodgman I think we need to defer this because it breaks interoperability tests with OpenSSL which currently still rely on this constant.

**Comment by ronald-cron-arm** [comment]
> 
> 
> @ronald-cron-arm Why was this removed from the Mbed TLS 3.0 Optional EPIC?

Because it is linked to the issue #4363 that is in the Mbed TLS 3.0 Optional EPIC and thus in the EPIC we can still see this PR through the issue.

**Comment by hanno-arm** [comment]
@ronald-cron-arm Got it!

**Review by daverodgman — DISMISSED** [review]
Looks good, comments have been addressed - thanks!

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've read the discussion in https://github.com/ARMmbed/mbedtls/issues/2188 and https://github.com/ARMmbed/mbedtls/issues/4363 and reviewed the proposal here.

I think the callback should have access to the SSL context. This was proposed in https://github.com/ARMmbed/mbedtls/issues/2188 and I think this is both simpler and more future-proof than the proposed alternatives such as identifying a connection by its random. I don't see any rationale for the current choice of parameters for the callback.

File: ChangeLog.d/key-export.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+API changes
+   * mbedtls_ssl_conf_export_keys_ext_cb() has been removed.
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply ChangeLog.d/key-export.txt:2]: Ok

File: include/mbedtls/ssl.h [hunk 4]
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
...
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_CLIENT_APPLICATION_TRAFFIC_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_SERVER_APPLICATION_TRAFFIC_SECRET,
+#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
+} mbedtls_ssl_key_export_type;
+
+/**
+ * \brief           Callback type: Export key alongside random values for
+ *                                 session identification, and PRF for
+ *                                 implementation of TLS key exporters.
+ *
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: I suggest to make the configuration SSL context specific instead, instead of passing the callback. This provides more flexibility to users.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: I don't like to add getters for the key material because then you have to explain if/when is it save to call them, e.g. what happens if you call them when inside/outside a handshake, or a renegotiation, ... It's much easier to directly pass the key material when it's ready.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:991]: I don't think making the callback _function_ connection-specific is particularly useful. On the other hand, making the callback _context_ would be.

> you have to explain if/when is it save to call them

For the purpose here it's enough to guarantee that they're available during the call to the export callback.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback? That doesn't make sense to me.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: > I don't think making the callback function connection-specific is particularly useful. On the other hand, making the callback context would be.

I'd be OK with making the callback function itself configuration-specific, but the callback context connection-specific.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:991]: > Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback?

Because if (when) we later discover that it's useful to access some other data in the callback, we can add a getter function. We can't add a parameter to the callback.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: Note that the new design allows to add more elements to the enum. And if that happens not be enough (unlikely, IMO), we can still add getters which users can access using a reference to the SSL context which is part of the callback context.

File: include/mbedtls/ssl.h [hunk 5]
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
...
+ * \param secret     The address of the buffer holding the secret
+ *                   that's being exporterd.
+ * \param secret_len The length of \p secret in bytes.
+ * \param client_random The client random bytes.
+ * \param server_random The server random bytes.
+ * \param tls_prf_type The identifier for the PRF used in the handshake
+ *                     to which the key belongs.
+ *
+ * \return          \c 0 if successful.
+ * \return          A negative error code on failure.
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:1003]: Good point, I think the callback should return void indeed.

File: include/mbedtls/ssl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:991]: I have made the callback and context connection-specific for now. I agree that the callback itself could be configuration specific, but I prefer to keep things together here.

File: ChangeLog.d/key-export.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply ChangeLog.d/key-export.txt]: Unless we're strict about this, I prefer keeping this under `API Changes`, because we've rather changed the export API than removed it.

**Comment by hanno-arm** [comment]
@mpg @gilles-peskine-arm @daverodgman I've made the key export callback and context connection specific, addressing #2188. 

Could you please review?

**Comment by gilles-peskine-arm** [comment]
You accidentally committed a modified `config.h`. Please amend the last commit to fix this.

File: include/mbedtls/ssl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:991]: I'm not convinced that it's better to pass all these arguments instead of letting the callback query them explicitly. But since the callback has access to the connection context, the API is extensible by adding getter functions, so the API is ok.

File: include/mbedtls/ssl.h [hunk 6]
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
...
+{
+   MBEDTLS_SSL_TLS_PRF_NONE,
+   MBEDTLS_SSL_TLS_PRF_SHA384,
+   MBEDTLS_SSL_TLS_PRF_SHA256,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA384,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA256
+}
+mbedtls_tls_prf_types;
+
+#if defined(MBEDTLS_SSL_EXPORT_KEYS)
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:973]: `MBEDTLS_SSL_EXPORT_KEYS` controls a very small amount of code, is not really a security risk (if you can corrupt the SSL context by adding a callback, you can surely corrupt the SSL context in many other ways), and has a wide variety of uses both for production and debugging. How about removing it (keeping it always on)? The fewer configuration options the better.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Ok for the design of the callback. One day we should unify the design of the many callbacks, but today is not the day.

* I think we should remove `MBEDTLS_SSL_EXPORT_KEYS` as a configuration option.
* https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r646175183 still needs adressing.
* The changelog entry could use some simplification.

File: include/mbedtls/ssl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/ssl.h:1003]: I think returning `void` is the right thing to do for this kind of callback. If anything goes wrong with the export for some reason, it doesn't invalidate anything in the connection - it might be a problem for the application, and then the application will decide what to do (including closing the connection if desired).

File: include/mbedtls/ssl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/ssl.h:991]: After reviewing the discussions in #2188, #4363 and here, my gut feeling is it's unlikely but not impossible that the callback will need access to more information in the future, and I think the current API matches that nicely: it's optimized for the likely case when the callback needs what it currently does, by passing that as parameters (avoiding the need for getter functions), but also support extensions with getters should they be needed.

File: ChangeLog.d/key-export.txt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+API changes
+   * mbedtls_ssl_conf_export_keys_ext_cb() and
+     mbedtls_ssl_conf_export_keys_cb() have been removed
+     and replaced by a new API
+     mbedtls_ssl_set_export_keys_cb().
+   * The signature of key export callbacks configured via
+     mbedtls_ssl_set_export_keys_cb() is different from that
+     of the previous mbedtls_ssl_conf_export_keys_cb(): First,
```
[the remark this answers is not in the document]
Comment (daverodgman) [reply ChangeLog.d/key-export.txt:8]: Probably a fair comment, but IMO it is reasonable as it stands (users can ignore the second bullet if they aren't interested) - not worth investing lots of time on perfecting the changelog entry.

File: include/mbedtls/ssl.h (hunk 6 above)
Comment (daverodgman) [reply include/mbedtls/ssl.h:973]: I think this is probably a useful simplification, the impact of having it always-on looks extremely small.

File: include/mbedtls/ssl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (daverodgman) [reply include/mbedtls/ssl.h:1003]: I think I agree with @mpg - not sure what Mbed TLS could usefully do with the error.

File: include/mbedtls/ssl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (daverodgman) [reply include/mbedtls/ssl.h:991]: If the answer might be "don't call it", should we document now that under some circumstances (that we don't need to define at this point), the callback may not be called?

File: include/mbedtls/ssl.h (hunk 6 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h:973]: I disagree here and would like to keep the option. Key export is for two use cases: (a) debugging, (b) application-specific key export. Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.) Finally, while I haven't measured it, multiple invocations from multiple places of an unknown function pointer, with multiple arguments each, should be enough code that you'd like to be able to remove it on a highly constrained system.

Long story short: No change here for now.

File: ChangeLog.d/key-export.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply ChangeLog.d/key-export.txt:8]: I'm OK swapping over to your wording.

File: include/mbedtls/ssl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:1003]: Changing to void.

[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h:1003]: For the record, I agree with @mpg's analysis.

**Comment by hanno-arm** [comment]
>I think we should remove MBEDTLS_SSL_EXPORT_KEYS as a configuration option.

No change for now, see comment.

> #4552 (comment) still needs adressing.

Fixed.

> The changelog entry could use some simplification.

Fixed.

@gilles-peskine-arm @daverodgman @mpg Could you please re-review?

**Review by daverodgman — DISMISSED** [review]
Looks good - I don't have a strong view on the MBEDTLS_SSL_EXPORT_KEYS question so I'm happy for this to land as-is (or with it removed if that ends up being the agreed approach).

**Comment by hanno-arm** [comment]
@mpg @gilles-peskine-arm As explained in https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r650030664, I'd suggest we take the opportunity of 3.0 to _disable_ this option by default. WDYT?

File: include/mbedtls/ssl.h (hunk 6 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:973]: > Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.)

I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

> multiple invocations from multiple places of an unknown function pointer, with multiple arguments each

There's a single call site, that's not the issue. For an M0+ build of this PR at 95d69cf9471bd37cd5ea3553a17349e322eb69fc (baremetal config):
```
  350331  350027 (TOTALS)
   19971   19927 ssl_cli.o
   25364   25224 ssl_msg.o
   20787   20767 ssl_srv.o
   20246   20146 ssl_tls.o
```
For comparison, on `development` at 57864faf847572fa7b4e752068be55ce50bf2e8e:
```
  367608  367452 (TOTALS)
   20755   20751 ssl_srv.o
   20226   20102 ssl_tls.o
    4417    4389 version_features.o
```
I must say I'm surprised that the difference is so much larger in this PR — ~300B vs ~150B. I don't understand why there's an impact outside `ssl_tls.o`.

We're only testing the build with `MBEDTLS_SSL_EXPORT_KEYS` disabled in some reference configurations. So it's not untested, but it's slightly fragile (we may be missing some common combinations).

----

So now I would like to understand the code size impact, whether or not the compile-time option remains.

Comment (hanno-arm) [reply include/mbedtls/ssl.h:973]: > I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

If we never make mistakes, you're right. But the mere fact that we're checking for a callback and passing a key to it increases our attack surface.

Comment (hanno-arm) [reply include/mbedtls/ssl.h:973]: > So now I would like to understand the code size impact, whether or not the compile-time option remains.

I think we're making better use of our time by agreeing whether this feature can be disabled by default.

Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:973]: The default build is the one meant to be used when linking the library dynamically. So there should be a strong reason to disable a feature. The security risk is barely perceptible (it's in TLS which uses plenty of other function pointers anyway), so not a reason to disable in dynamic builds.

Comment (hanno-arm) [reply include/mbedtls/ssl.h:973]: Fair point, if that's the established definition of the default config, I'm fine keeping it on by default.

It changes nothing w.r.t. my views to keep this configurable, though.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm argued that the default config is the one used for dynamic linkage. Users which obtain Mbed TLS as part of their distro should have the ability to make use of this feature, if only for debugging, thus we must keep it on by default. I withdraw my proposal to disable-by-default.

So there's no change pending from my side - I maintain the view that the feature should remain configurable.

File: include/mbedtls/ssl.h (hunk 6 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:973]: Explanation for the code size difference: it's due to short vs long field accesses in the context. After moving the export fields near the end of the context, I get:
```
   20222   20146 ssl_tls.o
```
So:

1. Please move the new fields towards the end of the context. It's a preexisting issue that the order of fields doesn't put often-used fields near the beginning, but let's not make it worse.
2. I still think we should remove `MBEDTLS_SSL_EXPORT_KEYS`. However, since this wasn't in the original scope of this pull request, and it can be done in 3.x since we'd just be making the feature always on, this is not a blocker here.

**Comment by mpg** [comment]
What's the status here? It looks to me like @hanno-arm has made the requested changes, but @gilles-peskine-arm labeled "needs: work" - is it only about the conflict that needs resolving, or do you also expect other changes related to previous feedback?

Also, I'm seeing two reviewers here already, so I'm removing myself as a reviewer; please protest if you think my review is needed here!

**Comment by gilles-peskine-arm** [comment]
After further investigation from my last round of review, I [requested to add the export fields later in the context structure](https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r650148282). I'm waiting for this to be done, and for CI to pass with no merge conflicts, before I re-review.

**Comment by hanno-arm** [comment]
Moved the fields and rebased on top of `development` to remove numerous trivial conflicts related to the introduction of `MBEDTLS_PRIVATE()`.

**Comment by hanno-arm** [comment]
@daverodgman @gilles-peskine-arm Can you please take another look?

**Review by daverodgman — APPROVED** [review]
This LGTM. The latest iteration changes an older commit (see git range-diff development 2369e3892 731870ad4 ) which seems to be fixing a mistake in the older version.

**Comment by gilles-peskine-arm** [comment]
CI is still pretty unhappy.

**Comment by gilles-peskine-arm** [comment]
CI's still unhappy.

**Comment by daverodgman** [comment]
Rebased onto latest development to try re-running the CI

**Comment by hanno-arm** [comment]
Yet another try...

**Review by daverodgman — APPROVED** [review]
LGTM

File: docs/3.0-migration-guide.d/key-export.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+This affects users of the SSL key export APIs:
+```
+    mbedtls_ssl_conf_export_keys_cb()
+    mbedtls_ssl_conf_export_keys_ext_cb()
+```
+
+Those APIs have been removed and replaced by the new API
+`mbedtls_ssl_set_export_keys_cb()`. This API differs from
+the previous key export API in the following ways:
+
```
Comment (gilles-peskine-arm) [inline docs/3.0-migration-guide.d/key-export.md:13]: Another way is that the new callback returns void, whereas the old callbacks returned an int value that was ignored. Since the old value was ignored, that's minor enough not to be mentioned in the changelog which is a high-level semantic overview, but I think it should be listed in the migration guide, because I'd expect the list here to be a complete list of steps to migrate existing code.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except for one omission in the migration guide.

File: docs/3.0-migration-guide.d/key-export.md (hunk 8 above)
Comment (daverodgman) [reply docs/3.0-migration-guide.d/key-export.md:13]: Fair point, I've discussed wording with @hanno-arm & pushed a commit with the addition to the migration guide.

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by daverodgman** [comment]
Sorry, typo in my first attempt.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by daverodgman — APPROVED** [review]
